akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala

```diff
@@ -49,26 +53,52 @@ private[http] object StreamUtils {
   }
 
   def captureTermination[T, Mat](source: Source[T, Mat]): (Source[T, Mat], Future[Unit]) = {
-    val promise = Promise[Unit]()
+    val (newSource, termination, _, _) = captureMaterializationTerminationAndKillSwitch(source)
+    (newSource, termination)
+  }
+  def captureMaterializationTerminationAndKillSwitch[T, Mat](source: Source[T, Mat]): (Source[T, Mat], Future[Unit], Future[Unit], KillSwitch) = {
+    val terminationPromise = Promise[Unit]()
+    val materializationPromise = Promise[Unit]()
+    val killResult = Promise[Unit]()
+    val killSwitch = new KillSwitch {
+      override def shutdown(): Unit = killResult.trySuccess(())
+      override def abort(ex: Throwable): Unit = killResult.tryFailure(ex)
+    }
     val transformer = new SimpleLinearGraphStage[T] {
       override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with InHandler with OutHandler {
+        override def preStart(): Unit = {
+          materializationPromise.trySuccess(())
```

Comment:
Hard to say if this might be 'too early'?

Reply:
Do you have some concrete issue in mind with that? What we use that for is for checking whether the stream has been materialized in time. So, for the pool it's enough to have any indication that any stage of the stream has been materialized.

Reply:
Agreed it's OK for the current use case and that's good enough since it's an internal API anyway :+1: